src/ray/object_manager/ownership_based_object_directory.cc

```diff
@@ -246,36 +283,55 @@ ray::Status OwnershipBasedObjectDirectory::UnsubscribeObjectLocations(
 ray::Status OwnershipBasedObjectDirectory::LookupLocations(
     const ObjectID &object_id, const rpc::Address &owner_address,
     const OnLocationsFound &callback) {
-  WorkerID worker_id = WorkerID::FromBinary(owner_address.worker_id());
-  std::shared_ptr<rpc::CoreWorkerClient> rpc_client = GetClient(owner_address);
-  if (rpc_client == nullptr) {
-    RAY_LOG(WARNING) << "Object " << object_id << " does not have owner. "
-                     << "LookupLocations returns an empty list of locations.";
-    io_service_.post([callback, object_id]() {
-      callback(object_id, std::unordered_set<NodeID>(), "", NodeID::Nil(), 0);
-    });
-    return Status::OK();
-  }
+  auto it = listeners_.find(object_id);
+  if (it != listeners_.end() && it->second.subscribed) {
+    // If we have locations cached due to a concurrent SubscribeObjectLocations
+    // call, and we have received at least one update from the owner about
+    // the object's creation, then call the callback immediately with the
+    // cached locations.
+    auto &locations = it->second.current_object_locations;
+    auto &spilled_url = it->second.spilled_url;
+    auto &spilled_node_id = it->second.spilled_node_id;
+    auto object_size = it->second.object_size;
+    io_service_.post(
+        [callback, object_id, locations, spilled_url, spilled_node_id, object_size]() {
+          callback(object_id, locations, spilled_url, spilled_node_id, object_size);
+        });
+  } else {
+    WorkerID worker_id = WorkerID::FromBinary(owner_address.worker_id());
+    std::shared_ptr<rpc::CoreWorkerClient> rpc_client = GetClient(owner_address);
+    if (rpc_client == nullptr) {
+      RAY_LOG(WARNING) << "Object " << object_id << " does not have owner. "
+                       << "LookupLocations returns an empty list of locations.";
+      io_service_.post([callback, object_id]() {
```

Comment:
Btw, is this post necessary?

Reply:
Yes, due to [this issue](https://github.com/ray-project/ray/issues/2959), further described in [this thread](https://github.com/ray-project/ray/pull/3227#discussion_r230614532).